src/main/java/dinkplugin/notifiers/LootNotifier.java

```diff
@@ -230,40 +231,51 @@ private void handleNotify(Collection<ItemStack> items, String dropper, LootRecor
                 NotificationBody.builder()
                     .text(notifyMessage)
                     .embeds(embeds)
-                    .extra(new LootNotificationData(serializedItems, dropper, type, kc, rarity))
+                    .extra(new LootNotificationData(augmentedItems, dropper, type, kc, rarity))
                     .type(NotificationType.LOOT)
                     .thumbnailUrl(ItemUtils.getItemImageUrl(keyItem.getId()))
                     .build()
             );
         }
     }
 
+    /**
+     * Converts {@link SerializedItemStack} loot to {@link RareItemStack}
+     *
+     * @param type the type of loot source
+     * @param source the name of the loot source
+     * @param reduced the looted items (after {@link ItemUtils#reduceItemStack(Iterable)} was performed)
+     * @return the dropped items augmented with rarity information (as available from the NPC drop table)
+     */
     @Nullable
-    private Map.Entry<SerializedItemStack, Double> getRarestDropRate(Collection<ItemStack> items, String dropper, LootRecordType type, Collection<Integer> deniedIds) {
+    private Collection<SerializedItemStack> getItemRarities(LootRecordType type, String source, Collection<SerializedItemStack> reduced) {
         if (type != LootRecordType.NPC) return null;
-        return items.stream()
-            .filter(item -> !deniedIds.contains(item.getId()))
+        return reduced.stream()
             .map(item -> {
-                OptionalDouble rarity = rarityService.getRarity(dropper, item.getId(), item.getQuantity());
-                if (rarity.isEmpty()) return null;
-                return Map.entry(item, rarity.getAsDouble());
-            })
-            .filter(Objects::nonNull)
-            .min(Comparator.comparingDouble(Map.Entry::getValue))
-            .map(pair -> {
-                ItemStack item = pair.getKey();
-                SerializedItemStack stack = ItemUtils.stackFromItem(itemManager, item.getId(), item.getQuantity());
-                return Map.entry(stack, pair.getValue());
+                OptionalDouble rarity = rarityService.getRarity(source, item.getId(), item.getQuantity());
+                return RareItemStack.of(item, rarity.isPresent() ? rarity.getAsDouble() : null);
             })
+            .collect(Collectors.toList());
+    }
+
+    @Nullable
+    private RareItemStack getRarestDropRate(Collection<SerializedItemStack> items, Collection<Integer> deniedIds) {
+        if (items == null) return null;
+        return items.stream()
+            .filter(RareItemStack.class::isInstance)
+            .map(RareItemStack.class::cast)
```

Comment:
can refactor this in a separate PR to avoid casting (and improve performance by an imperceptible amount)